text/0000-edition-2021-and-beyond.md

```diff
@@ -0,0 +1,194 @@
+* Code that compiles without a warning on Edition N should also
+  compile on Edition N + 1.
+* Migration between editions should generally be automated.
+* Editions make "skin-deep" changes, with all editions ultimately
+  compiling to a single common representation.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We use this section to try and convey the story that average users will need to understand.
+
+## What is a Rust edition?
+
+Every three years, we introduce a new Rust Edition. These editions are named after the year in which they occur, like Rust 2015 or Rust 2018. Each crate specifies the Rust edition that it requires in its `Cargo.toml` file via a setting like `edition = "2018"`. The purpose of editions is to give us a chance to introduce "opt-in" changes like new keywords that would otherwise have the potential to break existing code.
+
+When we introduce a new edition, we don't remove support for the older ones, so all crates continue to compile just as they ever did. Moreover, editions are fully interoperable, so there is no possibility of an "ecosystem split". This means that you can upgrade your crates to the new edition on whatever schedule works best for you.
+
+The release of a new edition is always a celebratory affair. It gives us a chance to look back at all the work that has gotten done over the last three years. The "opt-in" changes also allow us to introduce new features or syntax that would otherwise be impossible.
+
+## How do I upgrade between editions?
+
+Upgrading between editions is meant to be easy. The general rule is, if your code compiles without warnings, you should be able to opt into the new edition, and your code will compile.
+
+Along with each edition, we also release support for it in a tool called `rustfix`, which will automatically migate your code from the old edition to the new edition, preserving semantics along the way. You may have to do a bit of cleanup after the tool runs, but it shouldn't be much.
+
+## "Migrations" in an edition vs "idiom lints"
+
+When we release a new edition, it comes together with a certain set of "migrations". Migrations are the "breaking changes" introduced by the edition, except of course that since editions are opt-in, no code actually breaks. For example, if we introduce a new keyword, you will have to rename variables or functions using the old keyword, or else use Rust's `r#keyword` feature (which allows you to use a keyword as a regular variable/function/type name). As mentioned before, the edition comes with tooling that will make these changes for you, though sometimes you will want to cleanup the resulting code afterwards.
+
+In addition to those migrations, editions also come with a set of "idiom lints". These lints warn against deprecated patterns that we wish to discourage, even though they continue to work. Since these are lints, they don't cause your code to stop compiling. 
```

Comment:
I think this section could be worded to be more clear. When users think of "editions", they think of "the edition that my code is on", not "the highest edition that my toolchain supports". This section appears to be saying that if a user has code on Edition 2018, and they update to Rust 1.XX which supports Edition 2021, their Edition 2018 code will immediately begin to produce warnings for Edition 2021 features, without any form of opt-in being necessary from the user. I think this fact is obscured by language like "editions also come with"; there's either a conflation of editions with  toolchain versions that introduce editions, or it may be better served by emphasizing that when an edition is released, the previous edition can be mutated to add new warnings.

This is an important distinction to make because an entirely reasonable alternative formulation could be "when Rust 2021 is released, any code that opts-in to it will be subjected to warnings that will become errors in Rust 2024", which I don't think this section is proposing.

Reply:
Hmmm you're right that there are two possible interpretations. I'm curious what @steveklabnik thought it meant, but I thought it meant this:

* We will release rustc 1.XX, which is the first (stable) version to support Edition N.
* At this time, the edition lints will become warnings in Edition N-1 (or perhaps all editions before N, actually).
* For crates that opt into Edition N, those same lints are now errors.

In other words, you are being encouraged to adopt the new idioms, regardless of what Edition you are in. This is the same as most lints, which apply uniformly. The only real difference between editions is the "default lint level".

Reply:
Actually I think those edition lints could be added at any time, not necessarily *just* in rustc 1.XX.

Reply:
In other words, to complete my rambly set of comments, we basically just add new lints at any time, just like we always do -- but, along when we release a new edition, we can specify some set of lints whose defaults go from warn to error.

Reply:
Yes, I thought the same as you, Niko. Those bullets plus your amendments :)

Maybe it's worth adding a concrete example of this, to clear up confusion.

Reply:
Agreed.

Reply:
(Added to Planned Edits section)